### PR TITLE
Avoid error if proofer dbl-clicks some buttons

### PR DESCRIPTION
**Pull request by windymilla**
Buttons in the proofing interface that change the page's status can cause an error if double-clicked. Since first click changes the status, second click gives an error. Error only appears to occur on some systems.

Solution is to stop the second click from attempting to re-submit by returning false from onClick handler. On first click, set a user-defined data attribute to true, so second click knows it's the second click.

Affects WordCheck, Save as Done and Proofread Next Page (from proofing and WordCheck interfaces), Save as In Progress and Save as Done.

Testable at https://www.pgdp.org/~windymilla/c/

### CODE REVIEW

File: pinc/button_defs.inc
Comment (srjfoo) [inline pinc/button_defs.inc:4]: Duplicate include?

File: tools/proofers/dp_proof.js
Comment (srjfoo) [inline tools/proofers/dp_proof.js:244]: Is `btn.dataset.alreadysubmitted = true;` necessary, or could this part of the conditional just be `return true;`?

Comment (cpeel) [reply tools/proofers/dp_proof.js:244]: That's the flag we need to track if the button has been clicked already.

File: tools/proofers/spellcheck.inc
Comment (cpeel) [inline tools/proofers/spellcheck.inc:210]: Errant trailing comma.

**Review by cpeel — COMMENTED** [review]
Love this clever solution!!

File: tools/proofers/dp_proof.js
Comment (chrismiceli) [inline tools/proofers/dp_proof.js:239]: I would prefer we instead just mark the button disabled. That way for slow connections, there is an indication that something happened. Then we don't need to track anything since the 2nd click won't fire any event.

File: tools/proofers/dp_proof.js
Comment (chrismiceli) [inline tools/proofers/dp_proof.js:243]: per https://developer.mozilla.org/en-US/docs/web/api/htmlelement/dataset, all dataset properties are strings, to avoid any implicit conversions, I would set `btn.dataset.alreadysubmitted = 'true';`. Then we could do a strict equality above if we continued with this approach
`if (btn.dataset.alreadysubmitted === 'true') {`

File: tools/proofers/dp_proof.js
Comment (chrismiceli) [inline tools/proofers/dp_proof.js:239]: Can we add a unit test for this method easily?

File: pinc/button_defs.inc
Comment (windymilla) [reply pinc/button_defs.inc:4]: Thanks - will fix

File: tools/proofers/spellcheck.inc
Comment (windymilla) [reply tools/proofers/spellcheck.inc:210]: Thanks - copied from different context in `button_defs.inc` and not adjusted properly

File: tools/proofers/dp_proof.js
Comment (windymilla) [reply tools/proofers/dp_proof.js:239]: That's what I tried first :)
However, once you disable the button, the submission doesn't happen, even when you return true.

I had an intermediate version that changed the button label, and used that as the flag detected by the second click. 
I decided the hidden data attribute was better because
1. It doesn't need you to choose suitable text to change the label to (though "Submitting" would probably be suitable for most.
2. The new label text doesn't need translating
3. It looks a bit odd: even when you're on a fast connection, you see the button label text change, which is unusual behavior.
4. This code seems fairly safe for anything that might not support the `dataset`. The double click bug would still be able to happen of course, but I don't think it should break normal single clicking.

File: tools/proofers/dp_proof.js
Comment (windymilla) [reply tools/proofers/dp_proof.js:243]: Good idea, thanks - I'll do that

File: tools/proofers/dp_proof.js
Comment (windymilla) [reply tools/proofers/dp_proof.js:239]: I'm sorry, but I know nothing about that. Happy to do something if someone is able to explain what's needed

Comment (70ray) [reply tools/proofers/dp_proof.js:239]: I agree.

**Comment by windymilla** [comment]
I think I've addressed everything except the unit test suggestion - is such a test possible this method?

**Review by srjfoo — APPROVED** [review]
I was able to get really old versions of FF and not-so-old versions of Chrome (at least through 92) to fail on the main test server using browserstack. I have not tested all browsers on all OSs, just enough to verify that I could predictably get the error, and that this code fixes the problem.

File: tools/proofers/dp_proof.js
Comment (chrismiceli) [reply tools/proofers/dp_proof.js:239]: Essentially you need to add a file here:
https://github.com/DistributedProofreaders/dproofreaders/tree/master/SETUP/tests/jsTests
Then setup a file as described by
https://qunitjs.com/
You can open https://github.com/DistributedProofreaders/dproofreaders/blob/master/SETUP/tests/qunit.html to run the test. I imagine you'll need to create an <input element programmatically for the test.

Comment (chrismiceli) [reply tools/proofers/dp_proof.js:239]: I just think that if clicking the button does nothing, it should be disabled. I tried this and it seems to achieve what you want. We don't need to change the label in my opinion.
https://jsfiddle.net/yo42jueb/
I'm not completely opposed to what we have here, just think we can use browser features (disabled button) instead of implementing something ourselves.

Comment (windymilla) [reply tools/proofers/dp_proof.js:239]: Thanks @chrismiceli  - It may well be my naivety regarding the interaction between PHP, HTML and JS, and hence where to put everything, so I may have made the wrong edit, but my attempt to do what you suggested does not fix the bug, i.e. I still get the error if I double click. If I have done it correctly, I think the reason could be that by the time it's submitted, it's too late - the second click has already been made, which is why I had to use `onclick`, rather than `onsubmit`.  Maybe you can spot if I did something wrong in trying to copy the code from your jsfiddle:
I only tried it out on the WordCheck button, by commenting out the `onclick` I had previously added to `button_defs.inc`, adding your `submitForm` function to `dp_proof.js`, and adding `'onsubmit'="return submitForm()"` to the `<form...>` in `text_frame_std.inc`.
I'll try pushing a commit with those changes so you can see them in context and hopefully spot whether it's my error, or just that it doesn't trap the double click early enough.

**Comment by windymilla** [comment]
Note that as it currently stands, this code does not fix the bug - see earlier comments

File: tools/proofers/dp_proof.js
Comment (windymilla) [reply tools/proofers/dp_proof.js:239]: I'm afraid I have no idea at all about any of that - if you had time, would you be willing to guide me through the process at some point, maybe under a different PR @chrismiceli ? Thanks.

Comment (srjfoo) [reply tools/proofers/dp_proof.js:239]: > Thanks @chrismiceli - It may well be my naivety regarding the interaction between PHP, HTML and JS, and hence where to put everything, so I may have made the wrong edit, but my attempt to do what you suggested does not fix the bug, i.e. I still get the error if I double click. If I have done it correctly, I think the reason could be that by the time it's submitted, it's too late - the second click has already been made, which is why I had to use `onclick`, rather than `onsubmit`. Maybe you can spot if I did something wrong in trying to copy the code from your jsfiddle: I only tried it out on the WordCheck button, by commenting out the `onclick` I had previously added to `button_defs.inc`, adding your `submitForm` function to `dp_proof.js`, and adding `'onsubmit'="return submitForm()"` to the `<form...>` in `text_frame_std.inc`. I'll try pushing a commit with those changes so you can see them in context and hopefully spot whether it's my error, or just that it doesn't trap the double click early enough.

@chrismiceli -- I can confirm that the code currently in @windymilla's sandbox can still be made to raise the error associated by double-clicking on the WordCheck button.

Comment (chrismiceli) [reply tools/proofers/dp_proof.js:239]: Let me find time tomorrow to see what is going on with the disable option and maybe I can send a commit with the unit test change too if that is ok?

Comment (windymilla) [reply tools/proofers/dp_proof.js:239]: Thanks - that sounds like a good plan

Comment (srjfoo) [reply tools/proofers/dp_proof.js:239]: > I've sent a PR to your branch @windymilla here: [windymilla#1](https://github.com/windymilla/dproofreaders/pull/1) If you are ok with that, you can merge it and it should show up here. Sandbox here: https://www.pgdp.org/~chrismiceli/c.branch/form/

I've tested in @chrismiceli's sandbox, and can confirm that it (as far as I can tell) fixes the issue.

Something oddball that I experienced -- the project that I chose for my testing had French for the language. Now, on TEST, this doesn't mean a whole lot, and as the pages were actually in English, I eventually got tired of adding English as a dictionary to WC for the times it came up successfully. So I changed it to English. Whereupon I couldn't get the double-click to error, even in FF 52. (I tested in FF 52, and Chrome 70 and 90.)

It's actually pretty irrelevant, because regardless, it's a good idea to fix the issue. I mention it only because it's a good example of where speed (i.e. loading dictionaries) does seem to make a difference.

File: SETUP/tests/jsTests/dp_proofTests.js
Comment (windymilla) [inline SETUP/tests/jsTests/dp_proofTests.js:38]: Should this message say "button should be enabled"?

File: tools/proofers/text_frame_std.inc
Comment (windymilla) [inline tools/proofers/text_frame_std.inc:40]: Does `proof_frame_enh.inc` need this same edit?

**Review by windymilla — COMMENTED** [review]
Thanks @chrismiceli
That fix also seems to stop the error from happening. I agree it makes more sense to disable buttons than to ignore clicks programmatically. Also, if I understand correctly, in addition to double-clicking, your solution traps a person clicking on one button, then clicking on another button before the page refreshes.

I've added a couple of questions, but they may reflect more on my inexperience than any changes needed in the code.

File: SETUP/tests/jsTests/dp_proofTests.js
Comment (chrismiceli) [reply SETUP/tests/jsTests/dp_proofTests.js:38]: yup, that is a copy paste error

File: tools/proofers/text_frame_std.inc
Comment (chrismiceli) [reply tools/proofers/text_frame_std.inc:40]: yes I would imagine the change is very similar.

Comment (windymilla) [reply tools/proofers/text_frame_std.inc:40]: Thanks - I'll do those and push another commit

**Comment by windymilla** [comment]
Thanks @chrismiceli  - could you check if the edit I made to `submitForm` looks OK. It seems to fix double-clicking in the enhanced interface

File: tools/proofers/dp_proof.js
Comment (chrismiceli) [inline tools/proofers/dp_proof.js:243]: why do we need to disable images? do they have click handlers on them that submit the form?

Comment (windymilla) [reply tools/proofers/dp_proof.js:243]: That was my understanding from looking at the enhanced interface core. Without that, double clicking the wordcheck image button  causes the error we're trying to avoid.

Comment (srjfoo) [reply tools/proofers/dp_proof.js:243]: > That was my understanding from looking at the enhanced interface code. Without that, double clicking the wordcheck image button causes the error we're trying to avoid.

I'm glad you were able to verify this -- I couldn't get the error in the enhanced interface at all on browserstack, Windows or maOS, even using the oldest supported browsers. (And good that you remembered -- I was just testing in the standard interface.

Comment (chrismiceli) [reply tools/proofers/dp_proof.js:243]: Ah, checked it out, this is correct. input type=image behave like submit buttons per https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/image

**Review by chrismiceli — APPROVED** [review]

**Review by srjfoo — APPROVED** [review]

File: tools/proofers/text_frame_std.inc
Comment (70ray) [inline tools/proofers/text_frame_std.inc:26]: Rather than including dp_proof.js here would it be better to access the one already loaded in tools/proofers/proof.php by prepending "top."? as is done for initializeStuff() referring to dp_scroll. Similarly elsewhere.

**Review by 70ray — CHANGES_REQUESTED** [review]

File: tools/proofers/text_frame_std.inc
Comment (windymilla) [reply tools/proofers/text_frame_std.inc]: Thanks @70ray - another new thing I've learned :)

**Comment by windymilla** [comment]
There was another include in `spellcheck.inc` - fixed now. Thanks @srjfoo

**Review by cpeel — APPROVED** [review]
Thanks for the teamwork on this solution!

I am going to wait until after I cut the R202303 release (probably next week) before getting this merged in.

**Pull request merged**
